### PR TITLE
Lazy load Controllers

### PR DESCRIPTION
**Pull request by DonJayamanne**
Part of #7583

### CODE REVIEW

File: src/kernels/internalTypes.ts [hunk 1]
```diff
@@ -13,7 +13,6 @@ export enum ContributedKernelFinderKind {
 
 export interface IContributedKernelFinder<T extends KernelConnectionMetadata> extends IContributedKernelFinderInfo {
     kind: ContributedKernelFinderKind;
-    initialized: Promise<void>;
```
Comment (DonJayamanne) [inline src/kernels/internalTypes.ts:16]: Not required

File: src/kernels/jupyter/finder/remoteKernelFinder.ts [hunk 2]
```diff
@@ -71,13 +71,6 @@ export class RemoteKernelFinder implements IRemoteKernelFinder, IExtensionSingle
     private _onDidChangeKernels = new EventEmitter<void>();
     onDidChangeKernels: Event<void> = this._onDidChangeKernels.event;
 
-    private _initializeResolve: () => void;
-    private _initializedPromise: Promise<void>;
-
-    get initialized(): Promise<void> {
-        return this._initializedPromise;
```
Comment (DonJayamanne) [inline src/kernels/jupyter/finder/remoteKernelFinder.ts:78]: Now that we're push based (using events) this isn't required.

File: src/kernels/jupyter/types.ts [hunk 3]
```diff
@@ -360,8 +359,4 @@ export interface IServerConnectionType {
     onDidChange: Event<void>;
 }
 
-export interface IRemoteKernelFinder extends IContributedKernelFinder<RemoteKernelConnectionMetadata> {
-    listKernelsFromConnection(
```
Comment (DonJayamanne) [inline src/kernels/jupyter/types.ts:364]: This method was only used in tests, hence removed from the interface, not to be used in the extension code.
Was confusing why we had this when we're implemeneting an existing interface

File: src/kernels/types.ts [hunk 4]
```diff
@@ -611,7 +611,7 @@ export const IKernelFinder = Symbol('IKernelFinder');
 
 export interface IKernelFinder {
     onDidChangeKernels: Event<void>;
-    listKernels(cancelToken?: CancellationToken): Promise<KernelConnectionMetadata[]>;
+    kernels: KernelConnectionMetadata[];
```
Comment (DonJayamanne) [inline src/kernels/types.ts:614]: Push based API

File: src/notebooks/controllers/controllerLoader.ts [hunk 5]
```diff
@@ -70,12 +69,7 @@ export class ControllerLoader implements IControllerLoader, IExtensionSyncActiva
     public loadControllers(refresh?: boolean | undefined): Promise<void> {
         if (!this.controllersPromise || refresh) {
             const stopWatch = new StopWatch();
-            // Cancel previous load
-            if (this.loadCancellationToken) {
-                this.loadCancellationToken.cancel();
-            }
-            this.loadCancellationToken = new vscode.CancellationTokenSource();
-            this.controllersPromise = this.loadControllersImpl(this.loadCancellationToken.token)
+            this.controllersPromise = this.loadControllersImpl()
```
Comment (DonJayamanne) [inline src/notebooks/controllers/controllerLoader.ts:72]: Cancellation token is no longer useful as the API is not pull based anymore, hence nothing can be cancelled.
I.e. nothing is async

File: src/notebooks/controllers/controllerLoader.ts [hunk 6]
```diff
@@ -125,16 +119,11 @@ export class ControllerLoader implements IControllerLoader, IExtensionSyncActiva
         }
     }
 
-    private async loadControllersImpl(cancelToken: vscode.CancellationToken) {
+    private async loadControllersImpl() {
         // First off set our context to false as we are about to refresh
```
Comment (DonJayamanne) [inline src/notebooks/controllers/controllerLoader.ts:123]: To be discussed with @IanMatthewHuff 
THe only thing that makes this async is updateing a context variable, which doesn't seem to be valid anymore.
I think we can update the context variable without relying on controlers,
instead we can directly use Python API to check if tehre are any interpreters and update the context var accordingly, and that can be done in a seprate class, making this class cleaner

File: src/notebooks/controllers/controllerPreferredService.ts [hunk 7]
```diff
@@ -249,12 +249,15 @@ export class ControllerPreferredService implements IControllerPreferredService,
         let preferredConnection: KernelConnectionMetadata | undefined;
         const rankedConnections = await this.kernelRankHelper.rankKernels(
             uri,
+            this.registration.all,
```
Comment (DonJayamanne) [inline src/notebooks/controllers/controllerPreferredService.ts:252]: I don't think rankKernels should internall call the finder API, 
its not possible to rank kernels when we don't already have kernels, hence i don't think rank kernels should call any api to get the kernels, instead it should be given a list of kernsl to rank.
Else the method should be named `getRankedKernels`, and that complicates things.
I think this is simpler now (simpler to adopt the lazy loading of controllers)

File: src/notebooks/controllers/kernelRanking/kernelRankingHelper.ts [hunk 8]
```diff
@@ -19,36 +19,38 @@ import { IKernelRankingHelper } from '../types';
 @injectable()
 export class KernelRankingHelper implements IKernelRankingHelper {
     constructor(
-        @inject(IKernelFinder) private readonly kernelFinder: IKernelFinder,
```
Comment (DonJayamanne) [inline src/notebooks/controllers/kernelRanking/kernelRankingHelper.ts:22]: See comments earlier

**Review by IanMatthewHuff — APPROVED** [review]

**Pull request merged**